edgelet/aziot-edged/src/product_info.rs

```diff
@@ -0,0 +1,194 @@
+}
+
+impl ProductInfo {
+    pub fn try_load(p: impl AsRef<Path>) -> io::Result<Self> {
+        let bytes = fs::read(p)?;
+
+        toml::de::from_slice(&bytes).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+    }
+
+    pub fn from_system() -> io::Result<Self> {
+        let mut product = vec![];
+
+        // NOTE: operating system
+        product.push({
+            let b = io::BufReader::new(
+                fs::File::open("/etc/os-release")
+                    .or_else(|_| fs::File::open("/usr/lib/os-release"))?,
+            );
+
+            let mut name = None;
+            let mut version = None;
+            let mut comment = None;
+
+            for line in b.lines() {
+                match line?.trim().split_once("=") {
+                    Some(("ID", value)) => name = Some(strip_quotes(value)?.to_owned()),
+                    Some(("VERSION_ID", value)) => version = Some(strip_quotes(value)?.to_owned()),
+                    Some(("PRETTY_NAME", value)) => comment = Some(strip_quotes(value)?.to_owned()),
+                    _ => (),
+                }
+            }
+
+            Product {
+                name: name
+                    .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "os-release: ID"))?,
+                version: version.ok_or_else(|| {
+                    io::Error::new(io::ErrorKind::NotFound, "os-release: VERSION_ID")
+                })?,
+                comment,
+            }
+        });
+
+        // NOTE: kernel
+        product.push(Product {
+            name: Command::new("uname")
+                .arg("-s")
+                .output()
+                .and_then(read_output)?,
+            version: Command::new("uname")
+                .arg("-r")
+                .output()
+                .and_then(read_output)?,
+            comment: Some(
+                Command::new("uname")
+                    .arg("-v")
+                    .output()
+                    .and_then(read_output)?,
+            ),
+        });
+
+        // NOTE: device
+        product.push(Product {
+            name: read_utf8(fs::read("/sys/devices/virtual/dmi/id/product_name")?)?,
+            version: read_utf8(fs::read("/sys/devices/virtual/dmi/id/product_version")?)?,
+            comment: Some(read_utf8(fs::read(
+                "/sys/devices/virtual/dmi/id/sys_vendor",
+            )?)?),
+        });
+
+        Ok(Self { product })
+    }
+}
+
+impl fmt::Display for Product {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}/{}", self.name, self.version)?;
+
+        if let Some(comment) = &self.comment {
+            write!(f, " ({})", comment)?;
+        };
+
+        Ok(())
+    }
+}
+
+impl fmt::Display for ProductInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(first) = self.product.first() {
+            write!(f, "{}", first)?;
+
+            // NOTE: will not panic
```

Comment:
`let mut products = self.products.iter(); if let Some(first) = products.next() { ... for product in products { ... } }` will remove the need for the comment.

Or could just use `itertools::Itertools::format`